### ISSUE TITLE
[Epic: nwaku] PostgreSQL Optimisation phase 1

### ISSUE THREAD

**Issue by chair28980** [waku-org/nwaku#3098]
## Description

Perform optimizations suggested by @mrtrosen, Wings and @jakubgs (use of Redis; review pganalyze; look for horizontal scale solution.)

We may add other additional and small optimizations if needed.

**Comment by fryorcraken** [comment]
@Ivansete-status 

As discussed, it would be best to take a break and look at the actual performance we are getting before deciding to proceed with further steps such as Redis.

We know that down the line, we are planning orthogonal work that is going to help with SQL performance: Applying RLN, reducing message size, splitting traffic per shard, all the way to decentralizing store.

Hence, "good enough" performance is enough.

We have defined specific [FURPS](https://www.notion.so/Waku-FURPS-1498f96fb65c803faedef2a591c22c00?pvs=4#1498f96fb65c80bcadeec976db363cc3) for this:

> 1. Store message hash queries of less than 10 hashes are served (wire-to-wire) under 400 ms, as long as the database is less than 250 GB size and average message size under 500KiB (Vac-DST)
> 2. Store message time range queries of less than 24 hours, with less than 10 content topics, are served under 400 ms, as long as the database is less than 250 GB size and average message size under 500KiB (Vac-DST)

Next step would be to create Grafana panels on status prod and waku sandbox fleets to understand how far we are from this target.

I have created a dashboard for it [Waku Milestones Metrics / Direct Message Reliability](https://grafana.infra.status.im/d/D56HR5SNz/new-dashboard-waku-milestones-direct-message-reliability?orgId=1).

Let's add the following panels:

- Under "Store state [...]":
  - Database size
  - average message size
- Under "P1: [...]"
  - Wire to wire response time for hash queries on store nodes. With a visual threshold set to 400 ms
  - Number of hashes, in hash queries, with a visual threshold set to 10
-  Under "P2: [..]"
  - Wire to wire response time for time range queries on store nodes, with a visual threshold set to 400 ms
  - the length of the time range window for queries, with a visual threshold set to 24 hours



![Image](https://github.com/user-attachments/assets/f623f823-9792-4163-9f7b-afb40950908e)

**Comment by jm-clius** [comment]
I've previously had some success getting an idea of what messages are on which content topics, by performing SQL queries on https://superset.bi.status.im/sqllab/. I couldn't find this specific content topic (it may be too old or was never submitted for telemetry), but previous analyses found that these are often used for BACKUP or SYNC procedures. The work to split these functions into different shards will also help identify data hogging messages and procedures.

**Comment by fryorcraken** [comment]
Thanks @Ivansete-status for this analysis

> IMO, we need to carry on with investigation about the data stored on every contenttopic. I'd like to understand better what data is kept within those ~11.8k rows.

I am not sure I agree with this step. I am more keen to adapt the FURPS to the present reality of behaviour, and proceed with the other actions instead of further optimizing the DB for use cases we know are a poor usage of Waku.

I am specifically referring to:
- Applying RLN to cap traffic and hence DB usage
- Splitting traffic per shard, to better pin point this kind of issue and prioritize it in accordance to the underlying product feature (eg communities vs private chats)
- deploy SDS and further integration to partially replace time range queries with hash queries

**Comment by fryorcraken** [comment]
cc @plopezlpz to get some help here.

Can someone run Status desktop and check if they use `/waku/1/0xc49c91b7/rfc26`  ? might be easy to find out what this content topic is about.

**Comment by fryorcraken** [comment]
Also note the FURPS statement is:

> 🏗️ Store message time range queries of less than 24 hours, with less than 10 content topics, are served under 400 ms, as long as the database is less than 250 GB size and average message size under 500KiB (Vac-DST)

How big is the database? (could we add a panel of database size to https://grafana.infra.status.im/d/D56HR5SNz/direct-message-reliability?orgId=1&from=now-6h&to=now&timezone=utc ?)

Should we change the definition to replace "db of 250GB" by some measurement of number of messages?

**Comment by fryorcraken** [comment]
> Messages on that content topic are publishing in the rate of between 300 and 600 per hour. With payload size ranging between 700 and 1300 bytes.

Interestingly, this is not excessive for a single publisher in the context of the rate limitations we want to impose (eg 100 msgs per 10 minutes).

**Comment by fryorcraken** [comment]
Another question, is it possible to correlate the smart contract deployed for a community onchain, with the content topic used by the community? Do you see the hash of the community key on chain (which I think is what is used for content topic).

**Comment by Ivansete-status** [comment]
> Interestingly, this is not excessive for a single publisher in the context of the rate limitations we want to impose (eg 100 msgs per 10 minutes).

Yes, you are absolutely right, but the ~500 msgs/hour sounds weird for _Status_, IMO

**Comment by fryorcraken** [comment]
> > How big is the database?
> 
> 250GB
> 
> > Should we change the definition to replace "db of 250GB" by some measurement of number of messages?
> 
> Yes I think that would make more sense. For example, establishing a number of messages within 24h

Sounds good, let me know what number/what new statement we can commit to.

**Comment by fryorcraken** [comment]
@plopezlpz I do have more question:

> is it possible to correlate the smart contract deployed for a community onchain, with the content topic used by the community?



> [@fryorcraken](https://github.com/fryorcraken) , [@chair28980](https://github.com/chair28980) , [@plopezlpz](https://github.com/plopezlpz) - I suggest closing this issue because no further Postgres optimizations seem to be needed as all the queries are below 50ms for every page (100 messages.)

Sounds good, as long as panels are added.

**Comment by fryorcraken** [comment]
@Ivansete-status are we able to finish that soon?

**Comment by Ivansete-status** [comment]
@fryorcraken - this is blocked until we deploy a version >v0.34 in status fleet. Status fleet uses `v0.34.0` now.

The only missing task is to add some metric panels but those metrics are only available in >v0.34.

The `v0.35.1` cannot be deployed in `status.prod` because it apparently has a bug in filter. And we need to replicate that possible filter issue present in `v0.35.1`.

Nevertheless, this is not easy to replicated because `status-backend` cannot only deal with `status.staging` fleet as it eventually discovers peers from `status.prod`. Therefore, I need first to make some local manual temporary changes to status-go so that it doesn't discover nodes that don't run an incorrect version, and only interacts with nodes having the node.

Today Igor gave me some great hints on how to tackle that and I will proceed from there.

For more context:
- https://github.com/status-im/status-go/issues/6553

**Comment by fryorcraken** [comment]
<img width="966" height="629" alt="Image" src="https://github.com/user-attachments/assets/a0477b13-2053-4389-a2d1-e4f605220b7a" />


<img width="1008" height="636" alt="Image" src="https://github.com/user-attachments/assets/96d6d22d-80e6-4aec-ae04-52bb7ba86de8" />

<img width="949" height="618" alt="Image" src="https://github.com/user-attachments/assets/c932a034-1989-4b99-b9bb-198ba3fb02f1" />


We may need to change our FURPS to say that queries are done under 1s.

https://github.com/waku-org/pm/pull/338
